contracts/vortex/CarbonVortex.sol

```diff
@@ -138,109 +345,508 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
+                // withdraw vault token balance
+                _vault.withdrawFunds(token, payable(address(this)), vaultBalance);
+                totalFeeAmount += vaultBalance;
+            }
+            if (address(_oldVortex) != address(0)) {
+                // get old vortex token balance
+                uint256 oldVortexBalance = token.balanceOf(address(_oldVortex));
+                // withdraw old vortex token balance
+                _oldVortex.withdrawFunds(token, payable(address(this)), oldVortexBalance);
+                totalFeeAmount += oldVortexBalance;
+            }
+            feeAmounts[i] = totalFeeAmount;
+
             // get reward amount for token
-            rewardAmounts[i] = MathEx.mulDivF(balances[i], rewardsPercentage, PPM_RESOLUTION);
+            rewardAmounts[i] = MathEx.mulDivF(totalFeeAmount, rewardsPercentage, PPM_RESOLUTION);
         }
 
-        // convert tokens to BNT
-        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+        // go through all tokens and start / reset dutch auction if necessary
+        for (uint256 i = 0; i < tokens.length; ++i) {
             Token token = tokens[i];
-            // skip token if no token balance or token is BNT - no need to swap in this case
-            if (balances[i] == 0 || token == _bnt) {
+            uint256 totalFeeAmount = feeAmounts[i];
+            // skip token if no fees have accumulated or token pair is disabled
+            if (totalFeeAmount == 0 || _disabledPairs[token]) {
                 continue;
             }
-
-            // get trade amount for token
-            uint256 tradeAmount;
-            unchecked {
-                // safe because balances[i] >= rewardAmounts[i]
-                tradeAmount = balances[i] - rewardAmounts[i];
+            // skip the final target token
+            if (token == _finalTargetToken) {
+                continue;
+            }
+            // get fee and reward amounts
+            uint256 rewardAmount = rewardAmounts[i];
+            uint256 feeAmount = totalFeeAmount - rewardAmount;
+
+            if (token == _targetToken) {
+                // if _finalTargetToken is not set, directly transfer the fees to the transfer address
+                if (Token.unwrap(_finalTargetToken) == address(0)) {
+                    // safe due to nonReentrant modifier (forwards all gas fees in case of the native token)
+                    _targetToken.unsafeTransfer(_transferAddress, feeAmount);
+                    // increment totalCollected amount
+                    _totalCollected += feeAmount;
+                } else if (
+                    !_tradingEnabled(token) ||
+                    _amountAvailableForTrading(token) < _minTokenSaleAmounts[token] ||
+                    _auctionPriceIsBelowMinimum(token)
+                ) {
+                    // reset trading for target token
+                    _resetTradingTarget(rewardAmount);
+                }
+            } else {
```

Comment:
You can save the returned value of `_amountAvailableForTrading(token)` into a local variable here, instead of calling that function twice in some cases...

And unfortunately, zero times in other cases, i.e., when `!_tradingEnabled(token)`.

So a possible solution here is something like:
```
else if (!_tradingEnabled(token) {
    // reset trading for token
    _resetTrading(token, rewardAmount);
} else {
    uint128 amount = _amountAvailableForTrading(token);
    if (
        amount - feeAmount < _minTokenSaleAmounts[token] ||
        amount > _minTokenSaleAmountMultiplier * _minTokenSaleAmounts[token] ||
        _auctionPriceIsBelowMinimum(token)
    ) {
        // reset trading for token
        _resetTrading(token, rewardAmount);
    }
}
```

Reply:
fixed - extracted the _amountAvailableForTrading(token) in a separate variable. This means only the first auction (only once) for each token will be a bit more expensive, but all later resets are going to have 100 gas saved if the second function call is reached